File: backend/src/main/java/com/codeloom/backend/model/Problem.java
package com.codeloom.backend.model;

import java.time.Instant;
import java.util.List;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.relational.core.mapping.Column;
import org.springframework.data.relational.core.mapping.Table;

@Getter
@Table("problems")
@Builder(toBuilder = true)
@With
@AllArgsConstructor(access = AccessLevel.PACKAGE)
public class Problem {

    @Id
    @Column("problem_id")
    private Long id; // must be null for new entities so Spring Data JDBC does an INSERT

    @Column("slug")
    private String slug;

    @Column("title")
    private String title;

    @Builder.Default
    @Column("description")
    private String description = "";

    @Builder.Default
    @Column("difficulty")
    private ProblemDifficulty difficulty = ProblemDifficulty.EASY;

    @Column("constraints")
    private ProblemConstraints constraints;

    @Column("examples")
    private ProblemExamples examples;

    @Builder.Default
    @Column("hints")
    private List<String> hints = List.of();

    @CreatedDate
    @Column("created_at")
    private Instant createdAt;

    @LastModifiedDate
    @Column("updated_at")
    private Instant updatedAt;

    @Column("published_at")
    private Instant publishedAt;

    public boolean isDraft() {
        return publishedAt == null;
    }
}

File: backend/src/main/kotlin/com/codeloom/backend/model/Problem.kt
package com.codeloom.backend.model

import org.springframework.data.annotation.Id
import org.springframework.data.relational.core.mapping.Column
import org.springframework.data.relational.core.mapping.Table
import java.time.Instant

@Table("problems")
data class Problem(
    @Id
    @Column("problem_id")
    val id: Long? = null,
    @Column("slug")
    val slug: String,
    @Column("title")
    val title: String,
    @Column("description")
    val description: String = "",
    @Column("difficulty")
    val difficulty: ProblemDifficulty = ProblemDifficulty.EASY,
    @Column("constraints")
    val constraints: ProblemConstraints? = null,
    @Column("examples")
    val examples: ProblemExamples? = null,
    @Column("hints")
    val hints: Array<String> = arrayOf(),
    @Column("created_at")
    val createdAt: Instant = Instant.now(),
    @Column("updated_at")
    val updatedAt: Instant = Instant.now(),
    @Column("published_at")
    val publishedAt: Instant? = null,
)
